src/crypto/crypto_argon2.cc

```diff
@@ -0,0 +1,172 @@
+#ifndef OPENSSL_NO_ARGON2
+#include <openssl/core_names.h>
+
+using v8::FunctionCallbackInfo;
+using v8::JustVoid;
+using v8::Maybe;
+using v8::MaybeLocal;
+using v8::Nothing;
+using v8::Uint32;
+using v8::Value;
+
+Argon2Config::Argon2Config(Argon2Config&& other) noexcept
+    : mode{other.mode},
+      pass{std::move(other.pass)},
+      salt{std::move(other.salt)},
+      secret{std::move(other.secret)},
+      ad{std::move(other.ad)},
+      type{other.type},
+      iter{other.iter},
+      lanes{other.lanes},
+      memcost{other.memcost},
+      keylen{other.keylen} {}
+
+Argon2Config& Argon2Config::operator=(Argon2Config&& other) noexcept {
+  if (&other == this) return *this;
+  this->~Argon2Config();
+  return *new (this) Argon2Config(std::move(other));
+}
+
+void Argon2Config::MemoryInfo(MemoryTracker* tracker) const {
+  if (mode == kCryptoJobAsync) {
+    tracker->TrackFieldWithSize("pass", pass.size());
+    tracker->TrackFieldWithSize("salt", salt.size());
+    tracker->TrackFieldWithSize("secret", secret.size());
+    tracker->TrackFieldWithSize("ad", ad.size());
+  }
+}
+
+MaybeLocal<Value> Argon2Traits::EncodeOutput(Environment* env,
+                                             const Argon2Config& config,
+                                             ByteSource* out) {
+  return out->ToArrayBuffer(env);
+}
+
+Maybe<void> Argon2Traits::AdditionalConfig(
+    CryptoJobMode mode,
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int offset,
+    Argon2Config* config) {
+  Environment* env = Environment::GetCurrent(args);
+
+  config->mode = mode;
+
+  ArrayBufferOrViewContents<char> pass(args[offset]);
+  ArrayBufferOrViewContents<char> salt(args[offset + 1]);
+  ArrayBufferOrViewContents<char> secret(args[offset + 6]);
+  ArrayBufferOrViewContents<char> ad(args[offset + 7]);
+  Utf8Value type_(env->isolate(), args[offset + 8]);
+
+  if (!pass.CheckSizeInt32()) [[unlikely]] {
+    THROW_ERR_OUT_OF_RANGE(env, "pass is too large");
+    return Nothing<void>();
+  }
+
+  if (!salt.CheckSizeInt32()) [[unlikely]] {
+    THROW_ERR_OUT_OF_RANGE(env, "salt is too large");
+    return Nothing<void>();
+  }
+
+  if (!secret.CheckSizeInt32()) [[unlikely]] {
+    THROW_ERR_OUT_OF_RANGE(env, "secret is too large");
+    return Nothing<void>();
+  }
+
+  if (!ad.CheckSizeInt32()) [[unlikely]] {
+    THROW_ERR_OUT_OF_RANGE(env, "ad is too large");
+    return Nothing<void>();
+  }
+
+  const bool isAsync = mode == kCryptoJobAsync;
+  config->pass = isAsync ? pass.ToCopy() : pass.ToByteSource();
+  config->salt = isAsync ? salt.ToCopy() : salt.ToByteSource();
+  config->secret = isAsync ? secret.ToCopy() : secret.ToByteSource();
+  config->ad = isAsync ? ad.ToCopy() : ad.ToByteSource();
+
+  if (type_.ToStringView() == "argon2i") {
+    config->type = ncrypto::Argon2Type::ARGON2I;
+  } else if (type_.ToStringView() == "argon2d") {
+    config->type = ncrypto::Argon2Type::ARGON2D;
+  } else if (type_.ToStringView() == "argon2id") {
```

Comment:
Not a big deal, but better to do `type_.ToStringView()` only once. Can replace either with a new variable containing that value or use a switch

Reply:
Can't use a switch here. I think I can just do `type_ == "argon2x"` instead though, `Utf8Value` implements comparison against C string

Reply:
I locally made the requested changes but I'm having issues when hashing an empty string in async mode, I'm investigating why that's happening now and will push as soon as I figure out why 😓